sql: error on result type changes in prepared statements

madelynnblue opened the pull request:
Check the result types (excluding empty results and NULLs) against
what the initial prepared statement expected. This also moves all
error handling in prepared statements to a common place, unifying
more details of the pgwire and executor prepare paths.

Fixes #16062

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16089)
<!-- Reviewable:end -->

tamird commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/executor.go, line 556 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr3M9XEZ8-n0FBpTPc:-Kkr3M9XEZ8-n0FBpTPd:bzglhgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/executor.go#L556)):*
> ```Go
> 		if result.Rows != nil && !result.Columns.TypesEqual(stmt.Columns) {
> 			results.Close(session.Ctx())
> 			return StatementResults{}, pgerror.NewError(pgerror.CodeFeatureNotSupportedError,
> ```

is this really the right code? seems surprising.

---

*[pkg/sql/executor.go, line 1337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr30o4V2nBRIuAPdfO:-Kkr30o5_n6bNBCCdDue:b-4xb82j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/executor.go#L1337)):*
> ```Go
> 		}
> 		results, err := e.ExecutePreparedStatement(session, prepared, &parser.PlaceholderInfo{Values: qArgs, Types: prepared.SQLTypes})
> 		if err == nil && len(results.ResultList) == 1 {
> ```

is it actually possible for `len(results.ResultList) > 1`? in other words, why can't this be a "normal" error check?

---

*[pkg/sql/pgwire/pgwire_test.go, line 1621 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr2Z-jSndojep8-lUh:-Kkr2Z-jSndojep8-lUi:b-pgmd8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/pgwire/pgwire_test.go#L1621)):*
> ```Go
> 		t.Fatal(err)
> 	}
> ```

random newline

---

*[pkg/sql/pgwire/pgwire_test.go, line 1636 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr2T8zQKrmtu1cnzZT:-Kkr2T8zQKrmtu1cnzZU:b-33aec3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/pgwire/pgwire_test.go#L1636)):*
> ```Go
> 	}
> 	if _, err := stmt.Exec(); !testutils.IsError(err, "must not change result type") {
> 		t.Fatal("expected error")
> ```

include the error as %v (it might not be nil)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 556 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr3M9XEZ8-n0FBpTPc:-Kkr5Y18rXg_luR0pwtB:b-dmgcrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/executor.go#L556)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this really the right code? seems surprising.
</blockquote></details>

Yup:

```
postgres=# EXECUTE x;
ERROR:  0A000: cached plan must not change result type
LOCATION:  RevalidateCachedQuery, plancache.c:718
```

---

*[pkg/sql/executor.go, line 1337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr30o4V2nBRIuAPdfO:-Kkr5nYqViBGm7MNPvVa:bppn00u) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/executor.go#L1337)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is it actually possible for `len(results.ResultList) > 1`? in other words, why can't this be a "normal" error check?
</blockquote></details>

No, there's no way for a prepared statement to return more than one result set.

---

*[pkg/sql/pgwire/pgwire_test.go, line 1621 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr2Z-jSndojep8-lUh:-Kkr6GipSTlQyyDCwLXm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/pgwire/pgwire_test.go#L1621)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

random newline
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/pgwire_test.go, line 1636 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr2T8zQKrmtu1cnzZT:-Kkr67dsjqxp8Z_4kqct:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/pgwire/pgwire_test.go#L1636)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

include the error as %v (it might not be nil)
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/executor.go, line 1337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr30o4V2nBRIuAPdfO:-Kkr8AajEUOuQjtUmQPO:b-epevtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/executor.go#L1337)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

No, there's no way for a prepared statement to return more than one result set.
</blockquote></details>

So can this be a normal error check?

---

*[pkg/sql/pgwire/pgwire_test.go, line 1635 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr8JSjQaoyFU1Elgpb:-Kkr8JSkUK4F_5LdBH86:bxl5vys) ([raw file](https://github.com/cockroachdb/cockroach/blob/640c5e6eb150dd86bc11d3e3d48cb28f704c90f8/pkg/sql/pgwire/pgwire_test.go#L1635)):*
> ```Go
> 	}
> 	if _, err := stmt.Exec(); !testutils.IsError(err, "must not change result type") {
> 		t.Fatalf("expected error: %v", err)
> ```

"unexpected"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod existing comments

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-KkrARwC5fHAvSshdQIG:ba5p8ak)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr30o4V2nBRIuAPdfO:-KkrIsNNvtTdC9wDrWl8:b-yjvfxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfbd0ff80bc01b4bc5bca82d64308fdbffe2055/pkg/sql/executor.go#L1337)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

So can this be a normal error check?
</blockquote></details>

No, because it can also return 0 results, so we still have to check the length. Well, I used to have a normal error check, and a `return Result{}, nil` below, but that seemed redundant. I can change it back to that if you feel strong.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 1635 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr8JSjQaoyFU1Elgpb:-KksfrmPxuwpK4bvdtXZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/640c5e6eb150dd86bc11d3e3d48cb28f704c90f8/pkg/sql/pgwire/pgwire_test.go#L1635)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"unexpected"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

tamird commented:
I believe @knz had some comments about the correctness of this implementation, though I don't see them here.

Ah, they were made on the issue, rather than this PR. https://github.com/cockroachdb/cockroach/issues/16062#issuecomment-303726066

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-Kl-GzaZOBixCDNMRxDB:b-agoavh)*
<!-- Sent from Reviewable.io -->

mjibson commented:
@knz I've added a test that addresses your concern, I think, and it does indeed prevent the INSERT from running, so I think we're safe. Can you give a look?

knz commented on pkg/sql/executor.go:
```diff
@@ -1332,14 +1344,11 @@ func (e *Executor) execStmtInOpenTxn(
-		if len(results.ResultList) > 1 {
-			return Result{}, errWrongNumberOfPreparedStatements(len(results.ResultList))
+		results, err := e.ExecutePreparedStatement(session, prepared, &parser.PlaceholderInfo{Values: qArgs, Types: prepared.SQLTypes})
+		if err == nil && len(results.ResultList) == 1 {
```
The check for the number of results here is weird, if err is nil and the length of the results is >0, we lose the results and they are not Close()d either.

I think this would read better as
```
if err != nil{
   return Result{}, err
}
// optional check:
if len(results.ResultList) > 1 {
   panic(...)
}
return results.ResultList[0], nil
```

Also perhaps `ExecutePreparedStatement` could return a single result instead of a list?

knz reviewed (commented):
This code still does not properly fail.
As this is written, if (for example) an INSERT is prepared, the schema change is performed, and the INSERT is then executed, *the execute will be performed and the data will be inserted into the database* even though the code you have added subsequently reports an error to the client.

You should either find a way to ensure a failing check on the types **aborts the transaction** (which is not happening currently), or use a different approach with properly verifies the schema version (I think this would be more robust, and more importantly, far more correct).

mjibson commented:
I think one of us is missing something. I added a test that:

1. prepares an insert
2. executes a schema change
3. executes the prepared insert
4. observes that the insert did not take place

Or is the test an logic_test/prepare not doing that?

---

Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/executor.go, line 1348 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr1CPoGogIzh84NGXf-r4-1348:-KlPZhk_DUsi7aCXljLI:b-ec075b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6665ea1883ca586c08455c45c584d9367a3ca5/pkg/sql/executor.go#L1348)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The check for the number of results here is weird, if err is nil and the length of the results is >0, we lose the results and they are not Close()d either.

I think this would read better as
```
if err != nil{
   return Result{}, err
}
// optional check:
if len(results.ResultList) > 1 {
   panic(...)
}
return results.ResultList[0], nil
```

Also perhaps `ExecutePreparedStatement` could return a single result instead of a list?
</blockquote></details>

I've made some changes here to be similar to yours. I don't think it's a good idea to return a single result because 1) in the normal case of prepared statements being executed through pgwire, we will have to reform it back into a StatementResults, and 2) zero results is valid if we are executing the empty prepared statement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-KlPaI9n2Nwmhwbgix_u:b-la1583)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes correct, the logic test is insufficient to test this. 

The problem here is that the PREPARE and EXECUTE *statements* are executed via the Executor logic and thus inside the error check that aborts the transaction.

In contrast, a pgwire prepare message is processed outside of the executor logic, and there the Executor will not see the error and not prevent the txn from committing.

Try via pgwire_test.go to see what I mean.

(Note: also, since the code paths are separate, we need *both* tests. I am not suggesting you replace the logic test by a pgwire test entirely).

---

Reviewed 1 of 1 files at r5.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/executor.go, line 1348 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr1CPoGogIzh84NGXf-r4-1348:-KlPbIlLgDHj0Ahq4Rpk:bbgyd96) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6665ea1883ca586c08455c45c584d9367a3ca5/pkg/sql/executor.go#L1348)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I've made some changes here to be similar to yours. I don't think it's a good idea to return a single result because 1) in the normal case of prepared statements being executed through pgwire, we will have to reform it back into a StatementResults, and 2) zero results is valid if we are executing the empty prepared statement.
</blockquote></details>

Ack. I still think the code should contain an explicit check that there are no more results beyond `[0]`, that would need to be `Close()`d.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-KlPbcFcIP3iGL-oO8E-:b-ls5sqx)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I've changed this to instead plumb down the expected results into the executor. I don't like calling plan.prepare down there, but I'm not aware of a better way. It is possible calling that is not idempotent, but all the tests pass. Any thoughts about doing it this way? Ugly.

knz commented:
Reviewed 2 of 5 files at r1, 1 of 2 files at r4, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 595 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-OOJnQN8y5mpTpWj:-KlZ-OOKrHPeLioPgLxt:b4cnjhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L595)):*
> ```Go
> 	// for each result in the reply.
> 	results := e.execParsed(session, stmts, pinfo, copyMsgNone, []sqlbase.ResultColumns{stmt.Columns})
> 	if len(results.ResultList) > 1 {
> ```

Same problem with having this check here again -- if the error occurs, the txn is still committed.
The check needs to be deeper.

---

*[pkg/sql/executor.go, line 750 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-lIy813jOlPefGCe:-KlZ-lIz0SnVo0NwLLW-:bgwd8pr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L750)):*
> ```Go
> 		var expect []sqlbase.ResultColumns
> 		if expectResultColumns != nil {
> 			expect = expectResultColumns[:len(stmtsToExec)]
> ```

I'm confused: how is this logic supposed to work since there is at most a single statement to execute?

---

*[pkg/sql/executor.go, line 1419 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ0UHgeJK65QOEHZch:-KlZ0UHgeJK65QOEHZci:b-y66t23) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L1419)):*
> ```Go
> 
> 	if expectResultColumns != nil {
> 		plan, err := planner.prepare(session.Ctx(), stmt)
> ```

This is  not the right location to do this. The check must be present inside execStmtInParallel() and execStmt(). I strongly suggest factoring it inside `makePlan()`.

If I were you I would proceed as follows:
- define a new type `type Statement struct { ast parser.Statement; expectedTypes ResultColumns }`, and use this new `Statement` struct instead of `parser.Statement` everywhere in this package; this will take care of removing the extra arguments you just added and that indeed feel "ugly"
- make `makePlan()` use the `expectedTypes` field, as you intend

---

*[pkg/sql/sqlbase/result_columns.go, line 64 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlYzV5IjfL2KAcwohTI:-KlYzV5J-R37wy8JQNDD:b-uy7lu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/sqlbase/result_columns.go#L64)):*
> ```Go
> 	}
> 	for i, c := range r {
> 		if c.Typ == parser.TypeNull || other[i].Typ == parser.TypeNull {
> ```

Can you explain what this special casing with NULL is about. Also, it misses a test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: error on result type changes in prepared statements

Create a new sql.Statement struct to hold optional expected result
columns. These are checked in makePlan. This change thus triggers a
lot of renaming and repackaging since this new sql.Statement type is
now used in most places.

Fixes #16062

mjibson commented:
I've done a major refactor to create a new sql.Statement type so as to not pass down the result columns as a separate variable, as suggested by knz. RFAL.

knz commented:
This looks *much* better. So many thanks for introducing `sql.Statement`. It's great.
I still have two concerns remaining, see reviewable for the 2 comments still open.

---

Reviewed 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-KldVUJfsqFKJaa0fUKh:bgljppr)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1348 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr1CPoGogIzh84NGXf-r4-1348:-Kl_AdQiy0qkt5shTvbk:bc1w3x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6665ea1883ca586c08455c45c584d9367a3ca5/pkg/sql/executor.go#L1348)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ack. I still think the code should contain an explicit check that there are no more results beyond `[0]`, that would need to be `Close()`d.
</blockquote></details>

execPrepared will return an error and close results if there are more than 1. Do you think that logic needs to be duplicated here? I think the comment is clear enough.

---

*[pkg/sql/executor.go, line 595 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-OOJnQN8y5mpTpWj:-Kl_4-iKbgn538HtwlJj:b-srlhrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L595)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Same problem with having this check here again -- if the error occurs, the txn is still committed.
The check needs to be deeper.
</blockquote></details>

This is an error that should never occur because all prepared statements can only contain one statement during creation, so it's closer to a debug assertion.

---

*[pkg/sql/executor.go, line 750 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-lIy813jOlPefGCe:-Kl_A0a_nMVFLLMnFj5k:b-ycvs71) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L750)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm confused: how is this logic supposed to work since there is at most a single statement to execute?
</blockquote></details>

expectResultColumns is a slice of ResultColumns, and thus is used to specify either all or none of the statement results. Even though prepared statements are guaranteed to ever only have a single statement, I thought it was better to have this variable be a slice so that it could just mirror stmts. I'm not opposed to changing it, since it is kind of silly.

---

*[pkg/sql/executor.go, line 1419 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ0UHgeJK65QOEHZch:-Kle-Cx4-LAfduSeHkWs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L1419)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This is  not the right location to do this. The check must be present inside execStmtInParallel() and execStmt(). I strongly suggest factoring it inside `makePlan()`.

If I were you I would proceed as follows:
- define a new type `type Statement struct { ast parser.Statement; expectedTypes ResultColumns }`, and use this new `Statement` struct instead of `parser.Statement` everywhere in this package; this will take care of removing the extra arguments you just added and that indeed feel "ugly"
- make `makePlan()` use the `expectedTypes` field, as you intend



</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/result_columns.go, line 64 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlYzV5IjfL2KAcwohTI:-Kle-5tvE3g8xWs8cay3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/sqlbase/result_columns.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you explain what this special casing with NULL is about. Also, it misses a test.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
sql.Statement is looking :fire:! 

---

Review status: 13 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-Kle0DuoWn8eXbG6qn-W:bu9z0x2)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/executor.go, line 1348 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-Kkr1CPoGogIzh84NGXf-r4-1348:-Kle-yhiCecyvr9XttNP:bkmp1ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6665ea1883ca586c08455c45c584d9367a3ca5/pkg/sql/executor.go#L1348)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

execPrepared will return an error and close results if there are more than 1. Do you think that logic needs to be duplicated here? I think the comment is clear enough.
</blockquote></details>

No this comment was outdated.

---

*[pkg/sql/executor.go, line 595 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-OOJnQN8y5mpTpWj:-Kle-mVdMRxs8k8u3PjZ:b-2qzx0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L595)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

This is an error that should never occur because all prepared statements can only contain one statement during creation, so it's closer to a debug assertion.
</blockquote></details>

then the pgerror should be InternalError (which we'll soon catch with error reporting to the reg server)

---

*[pkg/sql/executor.go, line 750 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-lIy813jOlPefGCe:-Kle-uN-3nhQgYLqJAaQ:bsdt1sg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L750)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

expectResultColumns is a slice of ResultColumns, and thus is used to specify either all or none of the statement results. Even though prepared statements are guaranteed to ever only have a single statement, I thought it was better to have this variable be a slice so that it could just mirror stmts. I'm not opposed to changing it, since it is kind of silly.
</blockquote></details>

This is fine, thanks

---

*[pkg/sql/sqlbase/result_columns.go, line 64 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlYzV5IjfL2KAcwohTI:-Kle065r8ApvhlJsV7RP:b-711e0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/sqlbase/result_columns.go#L64)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can you dig this deeper then. I think it is possible that the PREPARE will resolve some non-NULL type, and then EXECUTE will infer TypeNull; however I believe the converse is impossible. This means only one of the two cases should be tested here.

Also I think this still misses a SQL prepare/execute test, not just a test of this function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 595 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlZ-OOJnQN8y5mpTpWj:-Kle6KuSXaK9Cegi-VET:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/executor.go#L595)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

then the pgerror should be InternalError (which we'll soon catch with error reporting to the reg server)
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/result_columns.go, line 64 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-KlYzV5IjfL2KAcwohTI:-Kle2A7DM-xhnVnbu_hr:b7kz53q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ef545897b39abdae141df2c5a3dbc8f9a9b22c/pkg/sql/sqlbase/result_columns.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you dig this deeper then. I think it is possible that the PREPARE will resolve some non-NULL type, and then EXECUTE will infer TypeNull; however I believe the converse is impossible. This means only one of the two cases should be tested here.

Also I think this still misses a SQL prepare/execute test, not just a test of this function.
</blockquote></details>

Yes, I had a small error and this has been changed so that only the RHS argument can be changed to NULL. There are tests in pgwire with SELECT CASE that cover the INT -> NULL case (which is how I found out I needed to write this block). Good?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089)*
<!-- Sent from Reviewable.io -->

knz commented:
terrific.
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16089#-:-Kle7j2nJjpwTIOcrv3y:ba3grhz)*
<!-- Sent from Reviewable.io -->

mberhault commented:
FYI: this has not been cherry-picked for `1.0.2` yet.

mjibson commented:
Yes, working on this today. Real bad merge.
